File: src/forge/local.py
"""Local spec source — reads issues from .forge/<name>/."""

from __future__ import annotations

import json
import re
from datetime import datetime
from pathlib import Path
from dataclasses import dataclass


@dataclass
class Issue:
    """A single issue from a local spec."""

    number: int
    filename: str
    path: Path
    content: str


class LocalSource:
    """Reads and manages issues from a local .forge/<name>/ spec directory."""

    def __init__(self, name: str) -> None:
        self.name = name
        self.spec_dir = Path(".forge") / name
        self._validate()

    def _validate(self) -> None:
        """Validate that the spec directory exists with prd.md and issues/."""
        if not self.spec_dir.is_dir():
            raise SystemExit(f"Error: spec directory not found: {self.spec_dir}/")
        if not (self.spec_dir / "prd.md").is_file():
            raise SystemExit(f"Error: {self.spec_dir}/prd.md not found")
        if not (self.spec_dir / "issues").is_dir():
            raise SystemExit(f"Error: {self.spec_dir}/issues/ not found")

    def _status_path(self) -> Path:
        return self.spec_dir / "status.json"

    def _read_status(self) -> set[int]:
        """Read completed issue numbers from status.json, initializing if needed."""
        path = self._status_path()
        if not path.exists():
            path.write_text('{ "completed": [] }\n')
            return set()
        data = json.loads(path.read_text())
        return set(data.get("completed", []))

    def _all_issues(self) -> list[tuple[int, str, Path]]:
        """Return all issue files as (number, filename, path) sorted by number."""
        issues_dir = self.spec_dir / "issues"
        results = []
        for f in sorted(issues_dir.iterdir()):
            if f.suffix == ".md":
                match = re.match(r"^(\d+)", f.name)
                if match:
                    results.append((int(match.group(1)), f.name, f))
        return results

    def get_prd_content(self) -> str:
        """Read and return the PRD content."""
        return (self.spec_dir / "prd.md").read_text()

    def get_next_issue(self) -> Issue | None:
        """Find the next incomplete issue, or None if all are complete."""
        completed = self._read_status()
        for number, filename, path in self._all_issues():
            if number not in completed:
                return Issue(
                    number=number,
                    filename=filename,
                    path=path,
                    content=path.read_text(),
                )
        return None

    def get_remaining_count(self) -> tuple[int, int]:
        """Return (remaining, total) issue counts."""
        completed = self._read_status()
        all_issues = self._all_issues()
        total = len(all_issues)
        remaining = sum(1 for num, _, _ in all_issues if num not in completed)
        return remaining, total

    def complete_issue(self, issue: Issue) -> None:
        """Mark an issue as complete: update status.json and append to progress.txt."""
        # Update status.json
        path = self._status_path()
        if path.exists():
            data = json.loads(path.read_text())
        else:
            data = {"completed": []}
        completed = set(data.get("completed", []))
        completed.add(issue.number)
        data["completed"] = sorted(completed)
        path.write_text(json.dumps(data, indent=2) + "\n")

        # Append to progress.txt
        progress_path = self.spec_dir / "progress.txt"
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M")
        with open(progress_path, "a") as f:
            f.write(f"{timestamp} — Issue {issue.number} complete: {issue.filename}\n")

File: src/local.ts
/** Local spec source — reads issues from .forge/<name>/. */

import { existsSync, readdirSync, readFileSync, writeFileSync, appendFileSync, statSync } from "fs";
import { join } from "path";
import type { Issue, Source } from "./types";

export class LocalSource implements Source<Issue> {
	readonly name: string;
	private readonly specDir: string;

	constructor(name: string) {
		this.name = name;
		this.specDir = join(".forge", name);
		this.validate();
	}

	private validate(): void {
		if (!existsSync(this.specDir) || !statSync(this.specDir).isDirectory()) {
			throw new Error(`Error: spec directory not found: ${this.specDir}/`);
		}
		if (
			!existsSync(join(this.specDir, "prd.md")) ||
			!statSync(join(this.specDir, "prd.md")).isFile()
		) {
			throw new Error(`Error: ${this.specDir}/prd.md not found`);
		}
		if (
			!existsSync(join(this.specDir, "issues")) ||
			!statSync(join(this.specDir, "issues")).isDirectory()
		) {
			throw new Error(`Error: ${this.specDir}/issues/ not found`);
		}
	}

	private statusPath(): string {
		return join(this.specDir, "status.json");
	}

	private readStatus(): Set<number> {
		const path = this.statusPath();
		if (!existsSync(path)) {
			writeFileSync(path, '{ "completed": [] }\n');
			return new Set();
		}
		const data = JSON.parse(readFileSync(path, "utf-8"));
		return new Set<number>(data.completed ?? []);
	}

	private allIssues(): Array<{ number: number; filename: string; path: string }> {
		const issuesDir = join(this.specDir, "issues");
		const entries = readdirSync(issuesDir).sort();
		const results: Array<{ number: number; filename: string; path: string }> = [];
		for (const name of entries) {
			if (!name.endsWith(".md")) continue;
			const match = name.match(/^(\d+)/);
			if (match) {
				results.push({
					number: parseInt(match[1], 10),
					filename: name,
					path: join(issuesDir, name),
				});
			}
		}
		return results;
	}

	async getPrdContent(): Promise<string> {
		return readFileSync(join(this.specDir, "prd.md"), "utf-8");
	}

	async getNextIssue(): Promise<Issue | null> {
		const completed = this.readStatus();
		for (const issue of this.allIssues()) {
			if (!completed.has(issue.number)) {
				return {
					number: issue.number,
					filename: issue.filename,
					path: issue.path,
					content: readFileSync(issue.path, "utf-8"),
				};
			}
		}
		return null;
	}

	async getRemainingCount(): Promise<[number, number]> {
		const completed = this.readStatus();
		const all = this.allIssues();
		const total = all.length;
		const remaining = all.filter((i) => !completed.has(i.number)).length;
		return [remaining, total];
	}

	async completeIssue(issue: Issue): Promise<void> {
		// Update status.json
		const path = this.statusPath();
		let data: { completed: number[] };
		if (existsSync(path)) {
			data = JSON.parse(readFileSync(path, "utf-8"));
		} else {
			data = { completed: [] };
		}
		const completed = new Set<number>(data.completed ?? []);
		completed.add(issue.number);
		data.completed = [...completed].sort((a, b) => a - b);
		writeFileSync(path, JSON.stringify(data, null, 2) + "\n");

		// Append to progress.txt
		const now = new Date();
		const timestamp = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")} ${String(now.getHours()).padStart(2, "0")}:${String(now.getMinutes()).padStart(2, "0")}`;
		appendFileSync(
			join(this.specDir, "progress.txt"),
			`${timestamp} — Issue ${issue.number} complete: ${issue.filename}\n`,
		);
	}
}
